libdnf-cli/utils/utf8.cpp

```diff
@@ -0,0 +1,164 @@
+    const char * ptr = str.c_str();
+    const char * end = ptr + strlen(ptr);
+
+    // multi-byte string state; required by mbrtowc()
+    std::mbstate_t state = std::mbstate_t();
+
+    // the wide char read from the input string
+    wchar_t wide_char = 0;
+
+    while (ptr < end) {
+        auto bytes = std::mbrtowc(&wide_char, ptr, MB_CUR_MAX, &state);
+        if (bytes <= 0) {
+            break;
+        }
+
+        // increase string width
+        result += wcwidth(wide_char);
+
+        // move the input string pointer by number of bytes read into the wide_char
+        ptr += bytes;
+    }
+
+    return result;
+}
+
+
+/// return an utf-8 sub-string that matches specified character count
+std::string substr_length(const std::string & str, std::size_t pos, std::size_t len) {
+    std::string result;
+
+    // pointers to the current position (defaults to begin) and the end of the input string
+    const char * ptr = str.c_str();
+    const char * end = ptr + strlen(ptr);
+    std::size_t index = 0;
+
+    // multi-byte string state; required by mbrtowc()
+    std::mbstate_t state = std::mbstate_t();
+
+    // the wide char read from the input string
+    wchar_t wide_char = 0;
+
+    while (ptr < end) {
+        auto bytes = std::mbrtowc(&wide_char, ptr, MB_CUR_MAX, &state);
+        if (bytes <= 0) {
+            break;
+        }
+
+        // skip first `pos` characters
+        if (pos > 0) {
+            ptr += bytes;
+            index += bytes;
+            pos--;
+            continue;
+        }
+
+        result += str.substr(index, bytes);
+        len--;
+
+        // move the input string pointer by number of bytes read into the wide_char
+        ptr += bytes;
+        index += bytes;
+
+        if (len == 0) {
+            break;
+        }
+    }
+
+    return result;
+}
+
+
+/// return an utf-8 sub-string that matches specified printable width
+std::string substr_width(const std::string & str, std::size_t pos, std::size_t wid) {
+    std::string result;
+
+    // pointers to the current position (defaults to begin) and the end of the input string
+    const char * ptr = str.c_str();
+    const char * end = ptr + strlen(ptr);
+    std::size_t index = 0;
+
+    // multi-byte string state; required by mbrtowc()
+    std::mbstate_t state = std::mbstate_t();
+
+    // the wide char read from the input string
+    wchar_t wide_char = 0;
+
+    while (ptr < end) {
+        auto bytes = std::mbrtowc(&wide_char, ptr, MB_CUR_MAX, &state);
+        if (bytes <= 0) {
+            break;
+        }
+
+        // skip first `pos` characters
+        if (pos > 0) {
+            ptr += bytes;
+            index += bytes;
+            pos--;
+            continue;
+        }
+
+        // increase string width
+        std::size_t char_width = wcwidth(wide_char);
+        if (char_width > wid) {
+            break;
+        }
+        result += str.substr(index, bytes);
```

Comment:
`result.append(ptr, bytes);`
and variable index is not needed.

Reply:
fixed